source/distributed/fullydistributed_tria.cc

```diff
@@ -0,0 +1,562 @@
+
+              dealii::Triangulation<dim, spacedim>::
+                execute_coarsening_and_refinement();
+            }
+
+
+
+          // 4a) set all cells artificial
+          for (auto cell = this->begin(); cell != this->end(); cell++)
+            {
+              if (cell->active())
+                cell->set_subdomain_id(
+                  dealii::numbers::artificial_subdomain_id);
+
+              if (settings & construct_multigrid_hierarchy)
+                cell->set_level_subdomain_id(
+                  dealii::numbers::artificial_subdomain_id);
+            }
+
+
+
+          // 4b) set actual (level_)subdomain_ids
+          for (unsigned int ref_counter = 0; ref_counter < parts.size();
+               ref_counter++)
+            {
+              auto cell      = this->begin(ref_counter);
+              auto cell_info = parts[ref_counter].begin();
+              for (; cell_info != parts[ref_counter].end(); ++cell_info)
+                {
+                  auto temp = cell_info->index;
+                  temp[0]   = coarse_cell_index_to_coarse_cell_id(temp[0]);
+                  while (temp != cell->id().template to_binary<dim>())
+                    cell++;
+
+                  if (cell->active())
+                    cell->set_subdomain_id(cell_info->subdomain_id);
+
+                  if (settings & construct_multigrid_hierarchy)
+                    cell->set_level_subdomain_id(cell_info->level_subdomain_id);
+                }
+            }
+        }
+
+      update_number_cache();
+    }
+
+
+
+    template <int dim, int spacedim>
+    void
+    Triangulation<dim, spacedim>::create_triangulation(
+      const std::vector<Point<spacedim>> &vertices,
+      const std::vector<CellData<dim>> &  cells,
+      const SubCellData &                 subcelldata)
+    {
+      AssertThrow(false, ExcMessage("Use the reinit-method instead!"));
+
+      dealii::parallel::Triangulation<dim, spacedim>::create_triangulation(
+        vertices, cells, subcelldata);
+    }
+
+
+
+    template <int dim, int spacedim>
+    Triangulation<dim, spacedim>::Triangulation(MPI_Comm mpi_communicaton,
+                                                Settings settings_)
+      : Triangulation<dim, spacedim>(mpi_communicaton,
+                                     mpi_communicaton,
+                                     settings_)
+    {}
+
+
+
+    template <int dim, int spacedim>
+    Triangulation<dim, spacedim>::Triangulation(
+      MPI_Comm mpi_communicator,
+      MPI_Comm mpi_communicator_coarse,
+      Settings settings_)
+      : parallel::DistributedTriangulationBase<dim, spacedim>(
+          mpi_communicator,
+          (settings_ & construct_multigrid_hierarchy) ?
+            static_cast<
+              typename dealii::Triangulation<dim, spacedim>::MeshSmoothing>(
+              dealii::Triangulation<dim>::none |
+              Triangulation<dim,
+                            spacedim>::limit_level_difference_at_vertices) :
```

Comment:
Is this necessary to create the correct mesh? I wonder if it works without...

Reply:
At the moment it is necessary since this is the way how I construct the `ConstructionData`. Is there a need to weaken this? Once we have hanging nodes and multigrid, we need it?

Reply:
One needs to satisfy this for GMG when reconstructing the local mesh from the oracle to have ghost level neighbors set correctly. My assumption is that you construct the p:fd:Tria from an object that already contains all cells and neighbors that you will need. Hence, I believe this flag is not needed for you here.

Reply:
As I said above, I have written the `ConstructionData` in such a way that it fulfills the condition`limit_level_difference_at_vertices` (by adding some artificial cells). I guess I can modify the definition of `ConstructionData` in such a way that we do not have to set this flag.